### PR TITLE
Refactor ViewController - extract logic

### PR DESCRIPTION
**Pull request by merecarvill**

### CODE REVIEW

File: TicTacToeApp/BoardButtons.swift [hunk 1]
```diff
@@ -0,0 +1,36 @@
+
+import UIKit
+
+public class BoardButtons {
```
Comment (ericmeyer) [inline TicTacToeApp/BoardButtons.swift:4]: I like this abstraction. It's a good abstraction on it's own and seems to be a good stepping stone towards a custom UI element if/when that is desired.

File: TicTacToeApp/PlayerMark.swift [hunk 2]
```diff
@@ -1,4 +1,6 @@
 
-public enum PlayerMark {
-  case X, O, NONE
+public enum PlayerMark: String {
+  case X = "X"
```
Comment (ericmeyer) [inline TicTacToeApp/PlayerMark.swift:3]: :+1: for splitting on multiple lines.

File: TicTacToeApp/GamePrompt.swift [hunk 3]
```diff
@@ -1,29 +1,28 @@
 
+import UIKit
+
 public class GamePrompt {
+  var prompt: UILabel
 
-  public init() {
-    // initialize with the UILabel?
+  public init(prompt: UILabel) {
+    self.prompt = prompt
```
Comment (ericmeyer) [inline TicTacToeApp/GamePrompt.swift:8]: I remember that this suggestion of injecting the `UILabel` was one we less certain on than others. Do you feel it improved the design?

File: TicTacToeApp/ViewController.swift [hunk 4]
```diff
@@ -1,152 +1,99 @@
...
-    presentViewController(confirmationDialogue, animated: true, completion: nil)
-  }
-
-  private func addAlertAction(alertController: UIAlertController, title: String, action: (UIAlertAction) -> Void) {
-    alertController.addAction(UIAlertAction(title: title, style: .Default, handler: action))
-  }
-
-  private func updateGamePrompt() {
-    infoLabel.text = gamePrompt.promptFor(game)
+  private func gameIsOver(game: Game) -> Bool {
```
Comment (ericmeyer) [inline TicTacToeApp/ViewController.swift:41]: This seems to be an SRP violation. The `ViewController` is knowing details of the rules of the game.

File: TicTacToeApp/ViewController.swift [hunk 5]
```diff
@@ -1,152 +1,99 @@
...
-// Refactoring alert dialog
-// Might be overkill. Let's revisit after board button refactorings
-//public class ConfirmResetGameAlert: UIAlertController {}
-// Alternatively, consider builder patter, or just wrapping the alert dialog
-//     i.e. composition over inheritance
-//public class ConfirmResetGameAlert {
-//  var alertController: UIAlertController?
-//}
-
 public class ViewController: UIViewController {
```
Comment (ericmeyer) [inline TicTacToeApp/ViewController.swift:4]: Even though `ViewController` is the generated name, it is conventional to name the class something more specific. For instance, what would happen if you needed to have a second controller?

File: TicTacToeAppTests/BoardButtonsSpec.swift [hunk 6]
```diff
@@ -0,0 +1,82 @@
+
+import Quick
+import Nimble
+import TicTacToeApp
+
+class BoardButtonsSpec: QuickSpec {
```
Comment (ericmeyer) [inline TicTacToeAppTests/BoardButtonsSpec.swift:6]: Is there now redundant test coverage between `BoardButtonsSpec` and `ViewControllerSpec`?

**Comment by ericmeyer** [comment]
Overall, I think this is a huge improvement to the `ViewController`, however it's still on the long side. I would look into making improvements regarding the `UIAlertController` code. This seems like a good candidate to wrap with a custom interface and inject a mock for testing.

File: TicTacToeApp/BoardButtons.swift [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+      button.setTitle(nil, forState: UIControlState.Normal)
+      button.enabled = true
+    }
+  }
+
+  public func disableInput() {
+    buttons.forEach{ $0.enabled = false }
+  }
+
+  private func getButtonBySpaceId(spaceId: Int) -> UIButton? {
```
Comment (ericmeyer) [inline TicTacToeApp/BoardButtons.swift:28]: What is `spaceId`?

File: TicTacToeApp/ViewController.swift [hunk 8]
```diff
@@ -1,152 +1,99 @@
...
-  private let gamePrompt = GamePrompt()
+  @IBOutlet public weak var prompt: UILabel!
+  @IBOutlet public var boardButtons: [UIButton]!
+  @IBOutlet public weak var resetButton: UIButton!
+  @IBOutlet var gameModeButtons: [UIButton]!
+
+  private var gameState = Game()
+  private var gameBoard: BoardButtons?
+  private var gamePrompt: GamePrompt?
   public var currentMode = GameMode.HumanVsHuman
```
Comment (ericmeyer) [inline TicTacToeApp/ViewController.swift:14]: Is seems a bit strange to have both a `gameState` and a `currentMode` variable. Could the game state encapsulate the mode?

File: TicTacToeApp/BoardButtons.swift (hunk 1 above)
Comment (merecarvill) [inline TicTacToeApp/BoardButtons.swift:4]: Thanks for pushing me in that direction.

File: TicTacToeApp/GamePrompt.swift (hunk 3 above)
Comment (merecarvill) [inline TicTacToeApp/GamePrompt.swift:8]: If subclassing UIView wasn't so frustrating, I'd have been tempted to implement GamePrompt as a subclass of UILabel. As it is, it seems about equivalent to me whether GamePrompt wraps a UILabel vs. outputs text which is assigned to a UILabel externally.

File: TicTacToeApp/ViewController.swift (hunk 4 above)
Comment (merecarvill) [inline TicTacToeApp/ViewController.swift:41]: Yeah, that's reasonable. My fix would be to extract that into Game#isOver() and query the game from ViewController. Sound good?

File: TicTacToeApp/ViewController.swift (hunk 5 above)
Comment (merecarvill) [inline TicTacToeApp/ViewController.swift:4]: That makes sense. Is it helpful to keep "View" in the name at all? We're thinking something like GameViewController or GameplayController. We'd certainly need to keep the names descriptive if we added a second ViewController. A second controller would probably help with naming by contrasting their functions.

File: TicTacToeApp/BoardButtons.swift (hunk 7 above)
Comment (merecarvill) [inline TicTacToeApp/BoardButtons.swift:28]: Same as the parameter labeled Id. A way of identifying spaces on the board. Per your previous comment I'm replacing all occurrences of Id and spaceId with "space"

File: TicTacToeApp/ViewController.swift (hunk 8 above)
Comment (merecarvill) [inline TicTacToeApp/ViewController.swift:14]: We've been thinking of doing that - we'll also definitely need to revisit GameMode in the future. We'll try to integrate it into Game when the time comes.

File: TicTacToeAppTests/BoardButtonsSpec.swift (hunk 6 above)
Comment (merecarvill) [inline TicTacToeAppTests/BoardButtonsSpec.swift:6]: There were a couple that were similar enough that we removed them. There are strong similarities between some tests of ViewController#resetGame() and some BoardButton tests - e.g. "re-enables board buttons when the game is reset" in ViewControllerSpec versus "enables all buttons when they are cleared of marks" in BoardButtons - but I think they're distinct enough to be worth keeping.

**Pull request merged**